### PR TITLE
feat(sysadvisor): list containers before updating

### PR DESCRIPTION
**Pull request by cheney-lin**
#### What type of PR is this?

Features

### CODE REVIEW

**Review by sun-yuliang — DISMISSED** [review]

File: pkg/agent/sysadvisor/plugin/qosaware/server/memory_server.go [hunk 1]
```diff
@@ -54,6 +61,85 @@ func (ms *memoryServer) RegisterAdvisorServer() {
...
+	}
+
+	if err := ms.StartListContainers(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func isUnimplementedError(err error) bool {
```
[the remark this answers is not in the document]
Comment (cheney-lin) [reply pkg/agent/sysadvisor/plugin/qosaware/server/memory_server.go]: done

File: pkg/agent/sysadvisor/plugin/qosaware/server/memory_server.go [hunk 2]
```diff
@@ -54,6 +61,85 @@ func (ms *memoryServer) RegisterAdvisorServer() {
 	ms.grpcServer = grpcServer
 }
 
+func (ms *memoryServer) Start() error {
+	if err := ms.baseServer.Start(); err != nil {
+		return err
+	}
+
+	if err := ms.StartListContainers(); err != nil {
```
[the remark this answers is not in the document]
Comment (cheney-lin) [reply pkg/agent/sysadvisor/plugin/qosaware/server/memory_server.go:68]: commented: Start is overrided

File: pkg/agent/sysadvisor/plugin/qosaware/server/memory_server.go [hunk 3]
```diff
@@ -54,6 +61,85 @@ func (ms *memoryServer) RegisterAdvisorServer() {
...
+	})
+	if err == nil {
+		return nil
+	}
+
+unimplementedError:
+	go func() {
+		general.Infof("wait %v to get add container query", durationToWait.String())
+		time.Sleep(durationToWait)
+		ms.sendCh <- types.TriggerInfo{TimeStamp: time.Now()}
```
[the remark this answers is not in the document]
Comment (cheney-lin) [reply pkg/agent/sysadvisor/plugin/qosaware/server/memory_server.go:101]: If listContainer is not implemented by QRM, we need to wait some time for adding container queries to populate metaCache.

File: pkg/agent/sysadvisor/plugin/qosaware/server/memory_server.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (waynepeking348) [reply pkg/agent/sysadvisor/plugin/qosaware/server/memory_server.go:68]: not for start but for StartListContainers

[the remark this answers is not in the document]
Comment (cheney-lin) [reply pkg/agent/sysadvisor/plugin/qosaware/server/memory_server.go:71]: To make sure metaCache is populated before memory advisor updating.

File: pkg/agent/sysadvisor/plugin/qosaware/server/memory_server.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (cheney-lin) [reply pkg/agent/sysadvisor/plugin/qosaware/server/memory_server.go:106]: If listonCtainers is not implemented by QRM, we need to wait `AddContainer` rpc to populate metaCache.

**Review by csfldf — APPROVED** [review]

**Review by waynepeking348 — APPROVED** [review]

**Pull request merged**